File: leetcode/1-100/24-swap-nodes-in-pairs/bench/iterative.c
/*
 * Benchmark workload — Swap Nodes in Pairs (LeetCode #24), iterative.
 * C mirror of bench/iterative.kara. Plain malloc/free linked list — Kāra's
 * `shared struct` and Rust's `Rc<RefCell<>>` are single-owner in this workload
 * (one list per iteration, swapped then freed at iteration end), so the C
 * mirror expresses the same alloc-per-node build + in-place pair re-link +
 * drop shape without RC overhead — the cleanest LLVM-backend floor.
 *
 * Same N/K and full-traversal sink. The swapped list is freed after the sink
 * read. See ../README.md § Benchmarks.
 */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

typedef struct ListNode {
    int64_t val;
    struct ListNode *next;
} ListNode;

static ListNode *node_new(int64_t val) {
    ListNode *p = (ListNode *)malloc(sizeof(ListNode));
    p->val = val;
    p->next = NULL;
    return p;
}

static void chain_free(ListNode *head) {
    while (head) {
        ListNode *next = head->next;
        free(head);
        head = next;
    }
}

static ListNode *swap_pairs(ListNode *head) {
    ListNode dummy;
    dummy.val = 0;
    dummy.next = head;
    ListNode *prev = &dummy;
    while (prev->next && prev->next->next) {
        ListNode *first = prev->next;
        ListNode *second = first->next;
        /* Re-link prev → second → first → rest. */
        first->next = second->next;
        second->next = first;
        prev->next = second;
        prev = first;
    }
    return dummy.next;
}

static ListNode *build_list(int64_t count) {
    if (count <= 0) return NULL;
    ListNode *head = node_new(1);
    ListNode *tail = head;
    for (int64_t v = 2; v <= count; v++) {
        ListNode *node = node_new(v);
        tail->next = node;
        tail = node;
    }
    return head;
}

static int64_t sum_list(ListNode *list) {
    int64_t s = 0;
    for (ListNode *c = list; c; c = c->next) s += c->val;
    return s;
}

int main(void) {
    const int64_t n_values = 100;
    const int64_t k_iters = 500000;

    int64_t total = 0;
    for (int64_t k = 0; k < k_iters; k++) {
        ListNode *list = build_list(n_values);
        ListNode *swapped = swap_pairs(list);
        total += sum_list(swapped);
        chain_free(swapped);
    }
    printf("%lld\n", (long long)total);
    return 0;
}

File: leetcode/1-100/24-swap-nodes-in-pairs/bench/iterative.rs
// Benchmark workload — Swap Nodes in Pairs (LeetCode #24), iterative.
// Rust mirror of bench/iterative.kara. Uses `Rc<RefCell<ListNode>>` to mirror
// Kāra's `shared struct` reference semantics — same per-node RC overhead and
// the same alloc-per-node build + in-place pair re-link shape. Same N/K and
// full-traversal sink. Compiled with `rustc -O`.
// See ../README.md § Benchmarks.

use std::cell::RefCell;
use std::rc::Rc;

type Link = Option<Rc<RefCell<ListNode>>>;

struct ListNode {
    val: i64,
    next: Link,
}

fn swap_pairs(head: Link) -> Link {
    let dummy = Rc::new(RefCell::new(ListNode { val: 0, next: head }));
    let mut prev = dummy.clone();
    loop {
        let first = prev.borrow().next.clone();
        let Some(first) = first else { break };
        let second = first.borrow().next.clone();
        let Some(second) = second else { break };
        // Re-link prev → second → first → rest.
        first.borrow_mut().next = second.borrow().next.clone();
        second.borrow_mut().next = Some(first.clone());
        prev.borrow_mut().next = Some(second);
        prev = first;
    }
    let result = dummy.borrow().next.clone();
    result
}

fn build_list(count: i64) -> Link {
    if count <= 0 {
        return None;
    }
    let head = Rc::new(RefCell::new(ListNode { val: 1, next: None }));
    let mut tail = head.clone();
    for v in 2..=count {
        let node = Rc::new(RefCell::new(ListNode { val: v, next: None }));
        tail.borrow_mut().next = Some(node.clone());
        tail = node;
    }
    Some(head)
}

fn sum_list(list: Link) -> i64 {
    let mut s = 0i64;
    let mut c = list;
    while let Some(n) = c {
        s += n.borrow().val;
        c = n.borrow().next.clone();
    }
    s
}

fn main() {
    let n_values: i64 = 100;
    let k_iters: i64 = 500_000;

    let mut total: i64 = 0;
    for _ in 0..k_iters {
        let list = build_list(n_values);
        let swapped = std::hint::black_box(swap_pairs(list));
        total += sum_list(swapped);
    }
    println!("{}", total);
}
